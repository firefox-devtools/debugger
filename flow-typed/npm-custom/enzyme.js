/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// Using this custom definition of the enzyme module avoids generating some
// flow errors that are unrelated to uses of enzyme itself...
// It would be nice to clean this up.
declare module "enzyme" {
  declare module.exports: any;
}
