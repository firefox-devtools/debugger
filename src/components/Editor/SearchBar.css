.search-bar {
  display: flex;
  flex-direction: column;
}

.search-bar .search-field {
  padding-left: 7px;
  height: var(--editor-searchbar-height);
}

.search-bar .close-btn {
  padding: 6px;
}

.search-bottom-bar * {
  -moz-user-select: none;
  user-select: none;
}

.search-bottom-bar {
  display: flex;
  flex-shrink: 0;
  justify-content: flex-end;
  width: calc(100% - 1px);
  height: var(--editor-second-searchbar-height);
  background-color: var(--theme-toolbar-background);
  border-bottom: 1px solid var(--theme-splitter-color);
  padding: 0 13px;
}

.search-bottom-bar button:focus {
  outline: none;
}

.search-bottom-bar .search-modifiers {
  display: flex;
  align-items: center;
}

.search-bottom-bar .search-modifiers button {
  padding: 0 3px;
  margin: 0 3px;
  border: none;
  background: none;
  width: 20px;
  height: 20px;
  border-radius: 3px;
}

.search-bottom-bar .search-modifiers button i {
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 0;
  width: 16px;
}

.search-bottom-bar .search-modifiers button svg {
  fill: var(--theme-comment-alt);
  height: 16px;
  width: 16px;
}

.search-bottom-bar .search-modifiers button:hover {
  background: var(--theme-toolbar-background-hover);
}

.search-bottom-bar .search-modifiers button:active {
  outline: none;
}

.search-bottom-bar .search-modifiers button.active svg {
  fill: var(--theme-selection-background);
}

.theme-dark .search-bottom-bar .search-modifiers button.active svg {
  fill: white;
}

.search-bottom-bar .search-type-toggles {
  display: flex;
  align-items: center;
  max-width: 68%;
}

.search-bottom-bar .search-type-name {
  padding: 1px 0 0 0;
  margin: 0 0 0 6px;
  border: none;
  background: transparent;
  color: var(--theme-comment-alt);
}

.search-bottom-bar .search-type-toggles .search-type-btn:active {
  outline: none;
}

.search-bottom-bar .search-type-toggles .search-type-btn.active {
  color: var(--theme-selection-background);
}

.theme-dark .search-bottom-bar .search-type-toggles .search-type-btn.active {
  color: white;
}

.search-bar .result-list {
  max-height: 230px;
}
