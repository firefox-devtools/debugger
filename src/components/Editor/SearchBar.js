// @flow

const React = require("react");
const { DOM: dom, PropTypes, createFactory } = React;
const { connect } = require("react-redux");
const { bindActionCreators } = require("redux");
const { findDOMNode } = require("react-dom");
const { isEnabled } = require("devtools-config");
const { filter } = require("fuzzaldrin-plus");
const Svg = require("../shared/Svg");
const actions = require("../../actions").default;
const {
  getFileSearchState,
  getFileSearchQueryState,
  getFileSearchModifierState
} = require("../../selectors");
const {
  find,
  findNext,
  findPrev,
  removeOverlay,
  countMatches,
  clearIndex
} = require("../../utils/editor");
const { getSymbols } = require("../../utils/parser");
const { scrollList } = require("../../utils/result-list");
const classnames = require("classnames");
const debounce = require("lodash/debounce");
const SearchInput = createFactory(require("../shared/SearchInput").default);
const ResultList = createFactory(require("../shared/ResultList").default);
const ImPropTypes = require("react-immutable-proptypes");

import type {
  FormattedSymbolDeclaration,
  SymbolDeclaration,
} from "../../utils/parser/utils";

function getShortcuts() {
  const searchAgainKey = L10N.getStr("sourceSearch.search.again.key");
  const fnSearchKey = L10N.getStr("symbolSearch.search.key");

  return {
    shiftSearchAgainShortcut: `CmdOrCtrl+Shift+${searchAgainKey}`,
    searchAgainShortcut: `CmdOrCtrl+${searchAgainKey}`,
    symbolSearchShortcut: `CmdOrCtrl+Shift+${fnSearchKey}`,
    searchShortcut: `CmdOrCtrl+${L10N.getStr("sourceSearch.search.key")}`
  };
}

type ToggleSymbolSearchOpts = {
  toggle: boolean,
  searchType: string
};

require("./SearchBar.css");

const SearchBar = React.createClass({
  propTypes: {
    editor: PropTypes.object,
    sourceText: ImPropTypes.map,
    selectSource: PropTypes.func.isRequired,
    selectedSource: ImPropTypes.map,
    searchOn: PropTypes.bool,
    toggleFileSearch: PropTypes.func.isRequired,
    searchResults: PropTypes.object.isRequired,
    modifiers: ImPropTypes.recordOf({
      caseSensitive: PropTypes.bool.isRequired,
      regexMatch: PropTypes.bool.isRequired,
      wholeWord: PropTypes.bool.isRequired
    }).isRequired,
    toggleFileSearchModifier: PropTypes.func.isRequired,
    query: PropTypes.string.isRequired,
    setFileSearchQuery: PropTypes.func.isRequired,
    updateSearchResults: PropTypes.func.isRequired
  },

  displayName: "SearchBar",

  getInitialState() {
    return {
      symbolSearchEnabled: false,
      selectedSymbolType: "functions",
      symbolSearchResults: [],
      selectedResultIndex: 0,
      count: 0,
      index: -1
    };
  },

  contextTypes: {
    shortcuts: PropTypes.object
  },

  componentWillUnmount() {
    const shortcuts = this.context.shortcuts;
    const {
      searchShortcut,
      searchAgainShortcut,
      shiftSearchAgainShortcut,
      symbolSearchShortcut
    } = getShortcuts();

    shortcuts.off(searchShortcut);
    shortcuts.off("Escape");
    shortcuts.off(searchAgainShortcut);
    shortcuts.off(shiftSearchAgainShortcut);
    shortcuts.off(symbolSearchShortcut);
  },

  componentDidMount() {
    // overwrite searchContents with a debounced version to reduce the
    // frequency of queries which improves perf on large files
    // $FlowIgnore
    this.searchContents = debounce(this.searchContents, 100);

    const shortcuts = this.context.shortcuts;
    const {
      searchShortcut,
      searchAgainShortcut,
      shiftSearchAgainShortcut,
      symbolSearchShortcut
    } = getShortcuts();

    shortcuts.on(searchShortcut, (_, e) => this.toggleSearch(e));
    shortcuts.on("Escape", (_, e) => this.onEscape(e));

    shortcuts.on(shiftSearchAgainShortcut, (_, e) =>
      this.traverseResults(e, true));

    shortcuts.on(searchAgainShortcut, (_, e) => this.traverseResults(e, false));

    if (isEnabled("symbolSearch")) {
      shortcuts.on(symbolSearchShortcut, (_, e) =>
        this.toggleSymbolSearch(e, {
          toggle: false,
          searchType: "functions"
        }));
    }
  },

  componentDidUpdate(prevProps: any, prevState: any) {
    const { sourceText, selectedSource, query, modifiers } = this.props;
    const searchInput = this.searchInput();

    if (searchInput) {
      searchInput.focus();
    }

    if (this.refs.resultList && this.refs.resultList.refs) {
      scrollList(this.refs.resultList.refs, this.state.selectedResultIndex);
    }

    const hasLoaded = sourceText && !sourceText.get("loading");
    const wasLoading = prevProps.sourceText &&
      prevProps.sourceText.get("loading");

    const doneLoading = wasLoading && hasLoaded;
    const changedFiles = selectedSource != prevProps.selectedSource &&
      hasLoaded;
    const modifiersUpdated = modifiers &&
      !modifiers.equals(prevProps.modifiers);

    const isOpen = this.props.searchOn || this.state.symbolSearchEnabled;
    const { selectedSymbolType, symbolSearchEnabled } = this.state;
    const changedSearchType = selectedSymbolType !=
      prevState.selectedSymbolType ||
      symbolSearchEnabled != prevState.symbolSearchEnabled;

    if (
      isOpen &&
      (doneLoading || changedFiles || modifiersUpdated || changedSearchType)
    ) {
      this.doSearch(query);
    }
  },

  onEscape(e: SyntheticKeyboardEvent) {
    this.closeSearch(e);
  },

  clearSearch() {
    const { editor: ed, query, modifiers } = this.props;
    if (ed && modifiers) {
      const ctx = { ed, cm: ed.codeMirror };
      removeOverlay(ctx, query, modifiers.toJS());
    }
  },

  closeSearch(e: SyntheticEvent) {
    const { editor: ed } = this.props;

    if (this.props.searchOn && ed) {
      this.clearSearch();
      this.props.toggleFileSearch(false);
      this.setState({
        symbolSearchEnabled: false,
        selectedSymbolType: "functions"
      });
      e.stopPropagation();
      e.preventDefault();
    }
  },

  toggleSearch(e: SyntheticKeyboardEvent) {
    e.stopPropagation();
    e.preventDefault();
    const { editor } = this.props;

    if (!this.props.searchOn) {
      this.props.toggleFileSearch();
    }

    if (this.state.symbolSearchEnabled) {
      this.clearSearch();
      this.setState({
        symbolSearchEnabled: false,
        selectedSymbolType: "functions"
      });
    }

    if (this.props.searchOn && editor) {
      const selection = editor.codeMirror.getSelection();
      this.setSearchValue(selection);
      if (selection !== "") {
        this.doSearch(selection);
      }
      this.selectSearchInput();
    }
  },

  toggleSymbolSearch(
    e: SyntheticKeyboardEvent,
    { toggle, searchType }: ToggleSymbolSearchOpts = {}
  ) {
    const { sourceText } = this.props;

    if (e) {
      e.preventDefault();
      e.stopPropagation();
    }

    if (!sourceText) {
      return;
    }

    if (!this.props.searchOn) {
      this.props.toggleFileSearch();
    }

    if (this.state.symbolSearchEnabled) {
      if (toggle) {
        this.setState({ symbolSearchEnabled: false });
      } else {
        this.setState({ selectedSymbolType: searchType });
      }
      return;
    }

    if (this.props.selectedSource) {
      this.clearSearch();
      this.setState({
        symbolSearchEnabled: true,
        selectedSymbolType: searchType
      });
    }
  },

  setSearchValue(value: string) {
    const searchInput = this.searchInput();
    if (value == "" || !searchInput) {
      return;
    }

    searchInput.value = value;
  },

  selectSearchInput() {
    const searchInput = this.searchInput();
    if (searchInput) {
      searchInput.setSelectionRange(0, searchInput.value.length);
    }
  },

  searchInput(): ?HTMLInputElement {
    const node = findDOMNode(this);
    if (node instanceof HTMLElement) {
      const input = node.querySelector("input");
      if (input instanceof HTMLInputElement) {
        return input;
      }
    }
    return null;
  },

  async updateSymbolSearchResults(query: string) {
    const {
      sourceText,
      updateSearchResults
    } = this.props;
    const { selectedSymbolType } = this.state;

    if (query == "" || !sourceText) {
      return;
    }

    const symbolDeclarations = await getSymbols(sourceText.toJS());
    const symbolSearchResults = filter(
      symbolDeclarations[selectedSymbolType],
      query,
      { key: "value" }
    );

    updateSearchResults({ count: symbolSearchResults.length });
    return this.setState({ symbolSearchResults });
  },

  async doSearch(query: string) {
    const {
      sourceText,
      setFileSearchQuery,
      editor: ed
    } = this.props;
    if (!sourceText || !sourceText.get("text")) {
      return;
    }

    setFileSearchQuery(query);

    if (this.state.symbolSearchEnabled) {
      return await this.updateSymbolSearchResults(query);
    } else if (ed) {
      this.searchContents(query);
    }
  },

  searchContents(query: string) {
    const {
      sourceText,
      modifiers,
      editor: ed,
      searchResults: { index }
    } = this.props;

    if (!ed || !sourceText || !sourceText.get("text") || !modifiers) {
      return;
    }

    const ctx = { ed, cm: ed.codeMirror };

    const newCount = countMatches(
      query,
      sourceText.get("text"),
      modifiers.toJS()
    );

    if (index == -1) {
      clearIndex(ctx, query, modifiers.toJS());
    }

    const newIndex = find(ctx, query, true, modifiers.toJS());
    this.props.updateSearchResults({
      count: newCount,
      index: newIndex
    });
  },

  traverseResults(e: SyntheticEvent, rev: boolean) {
    e.stopPropagation();
    e.preventDefault();

    const ed = this.props.editor;

    if (!ed) {
      return;
    }

    const ctx = { ed, cm: ed.codeMirror };

    const {
      query,
      modifiers,
      updateSearchResults,
      searchResults: { count, index }
    } = this.props;

    if (query === "") {
      this.props.toggleFileSearch(true);
    }

    if (index == -1 && modifiers) {
      clearIndex(ctx, query, modifiers.toJS());
    }

    if (modifiers) {
      const findFnc = rev ? findPrev : findNext;
      const newIndex = findFnc(ctx, query, true, modifiers.toJS());
      updateSearchResults({
        index: newIndex,
        count
      });
    }
  },

  // Handlers
  selectResultItem(e: SyntheticEvent, item: SymbolDeclaration) {
    const { selectSource, selectedSource } = this.props;

    if (selectedSource) {
      selectSource(selectedSource.get("id"), {
        line: item.location.start.line
      });

      this.closeSearch(e);
    }
  },

  onSelectResultItem(item: FormattedSymbolDeclaration) {
    const { selectSource, selectedSource } = this.props;
    if (selectedSource) {
      selectSource(selectedSource.get("id"), {
        line: item.location.start.line
      });
    }
  },

  async onChange(e: any) {
    return this.doSearch(e.target.value);
  },

  onKeyUp(e: SyntheticKeyboardEvent) {
    if (e.key != "Enter") {
      return;
    }

    this.traverseResults(e, e.shiftKey);
  },

  onKeyDown(e: SyntheticKeyboardEvent) {
    if (!this.state.symbolSearchEnabled || this.props.query == "") {
      return;
    }

    const searchResults = this.state.symbolSearchResults,
      resultCount = searchResults.length;

    if (e.key === "ArrowUp") {
      const selectedResultIndex = Math.max(
        0,
        this.state.selectedResultIndex - 1
      );
      this.setState({ selectedResultIndex });
      this.onSelectResultItem(searchResults[selectedResultIndex]);
      e.preventDefault();
    } else if (e.key === "ArrowDown") {
      const selectedResultIndex = Math.min(
        resultCount - 1,
        this.state.selectedResultIndex + 1
      );
      this.setState({ selectedResultIndex });
      this.onSelectResultItem(searchResults[selectedResultIndex]);
      e.preventDefault();
    } else if (e.key === "Enter") {
      if (searchResults.length) {
        this.selectResultItem(e, searchResults[this.state.selectedResultIndex]);
      }
      this.closeSearch(e);
      e.preventDefault();
    } else if (e.key === "Tab") {
      this.closeSearch(e);
      e.preventDefault();
    }
  },

  // Renderers
  buildSummaryMsg() {
    if (this.state.symbolSearchEnabled) {
      return L10N.getFormatStr(
        "sourceSearch.resultsSummary1",
        this.state.symbolSearchResults.length
      );
    }
    const { searchResults: { count, index }, query } = this.props;

    if (query.trim() == "") {
      return "";
    }

    if (count == 0) {
      return L10N.getStr("editor.noResults");
    }

    if (index == -1) {
      return L10N.getFormatStr("sourceSearch.resultsSummary1", count);
    }

    return L10N.getFormatStr("editor.searchResults", index + 1, count);
  },

  buildPlaceHolder() {
    const { symbolSearchEnabled, selectedSymbolType } = this.state;
    if (symbolSearchEnabled) {
      return L10N.getFormatStr(
        `symbolSearch.search.${selectedSymbolType}Placeholder`
      );
    }

    return L10N.getStr("sourceSearch.search.placeholder");
  },

  renderSearchModifiers() {
    if (!isEnabled("searchModifiers")) {
      return;
    }

    const {
      modifiers,
      toggleFileSearchModifier
    } = this.props;
    const { symbolSearchEnabled } = this.state;

    function searchModBtn(modVal, className, svgName, tooltip) {
      return dom.button(
        {
          className: classnames(className, {
            active: !symbolSearchEnabled && modifiers && modifiers.get(modVal),
            disabled: symbolSearchEnabled
          }),
          onClick: () =>
            !symbolSearchEnabled ? toggleFileSearchModifier(modVal) : null,
          title: tooltip
        },
        Svg(svgName)
      );
    }

    return dom.div(
      { className: "search-modifiers" },
      searchModBtn(
        "regexMatch",
        "regex-match-btn",
        "regex-match",
        L10N.getStr("symbolSearch.searchModifier.regex")
      ),
      searchModBtn(
        "caseSensitive",
        "case-sensitive-btn",
        "case-match",
        L10N.getStr("symbolSearch.searchModifier.caseSensitive")
      ),
      searchModBtn(
        "wholeWord",
        "whole-word-btn",
        "whole-word-match",
        L10N.getStr("symbolSearch.searchModifier.wholeWord")
      )
    );
  },

  renderSearchTypeToggle() {
    if (!isEnabled("symbolSearch")) {
      return;
    }
    const { toggleSymbolSearch } = this;
    const { symbolSearchEnabled, selectedSymbolType } = this.state;

    function searchTypeBtn(searchType) {
      return dom.button(
        {
          className: classnames("search-type-btn", {
            active: symbolSearchEnabled && selectedSymbolType == searchType
          }),
          onClick: e => {
            if (selectedSymbolType == searchType) {
              toggleSymbolSearch(e, { toggle: true, searchType });
              return;
            }
            toggleSymbolSearch(e, { toggle: false, searchType });
          }
        },
        searchType
      );
    }

    return dom.section(
      { className: "search-type-toggles" },
      dom.h1({ className: "search-toggle-title" }, "Search for:"),
      searchTypeBtn("functions"),
      searchTypeBtn("variables")
    );
  },

  renderBottomBar() {
    if (!isEnabled("searchModifiers") || !isEnabled("symbolSearch")) {
      return;
    }

    return dom.div(
      { className: "search-bottom-bar" },
      this.renderSearchTypeToggle(),
      this.renderSearchModifiers()
    );
  },

  renderResults() {
    const {
      symbolSearchEnabled,
      symbolSearchResults,
      selectedResultIndex
    } = this.state;
    const { query } = this.props;
    if (query == "" || !symbolSearchEnabled || !symbolSearchResults.length) {
      return;
    }

    return ResultList({
      items: symbolSearchResults,
      selected: selectedResultIndex,
      selectItem: this.selectResultItem,
      ref: "resultList"
    });
  },

  render() {
    const {
      searchResults: { count },
      query
    } = this.props;

    if (!this.props.searchOn) {
      return dom.div();
    }

    return dom.div(
      { className: "search-bar" },
      SearchInput({
        query,
        count,
        placeholder: this.buildPlaceHolder(),
        summaryMsg: this.buildSummaryMsg(),
        onChange: this.onChange,
        onKeyUp: this.onKeyUp,
        onKeyDown: this.onKeyDown,
        handleClose: this.closeSearch
      }),
      this.renderResults(),
      this.renderBottomBar()
    );
  }
});

module.exports = connect(
  state => {
    return {
      searchOn: getFileSearchState(state),
      query: getFileSearchQueryState(state),
      modifiers: getFileSearchModifierState(state)
    };
  },
  dispatch => bindActionCreators(actions, dispatch)
)(SearchBar);
