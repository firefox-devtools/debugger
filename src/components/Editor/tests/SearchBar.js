import { createFactory } from "react";
import { shallow } from "enzyme";
import SearchBar from "../SearchBar";

const SearchBarComponent = createFactory(SearchBar.WrappedComponent);

function generateDefaults() {
  return {
    query: "",
    searchOn: true,
    symbolSearchOn: true,
    searchResults: {},
    selectedSymbolType: "functions",
    symbolSearchResults: [],
    selectedResultIndex: 0
  };
}

function render(overrides = {}) {
  const defaults = generateDefaults();
  const props = { ...defaults, ...overrides };
  const component = shallow(new SearchBarComponent(props));
  return { component, props };
}

describe("SearchBar", () => {
  it("should render", () => {
    const { component } = render();
    expect(component).toMatchSnapshot();
  });
});
