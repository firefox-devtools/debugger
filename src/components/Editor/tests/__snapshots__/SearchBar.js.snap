// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should render 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="search-bar"
>
    <SearchInput
        count={undefined}
        handleClose={[Function]}
        handleNext={[Function]}
        handlePrev={[Function]}
        onChange={[Function]}
        onKeyUp={[Function]}
        placeholder="Search in file…"
        query=""
        size=""
        summaryMsg=""
    />
    <div
        className="search-bottom-bar"
    >
        <div
            className="search-modifiers"
        >
            <button
                className="regex-match-btn"
                onClick={[Function]}
                title="Regex"
            >
                <InlineSVG
                    className="regex-match"
                    element="i"
                    raw={false}
                    src="<svg></svg>"
                />
            </button>
            <button
                className="case-sensitive-btn"
                onClick={[Function]}
                title="Case sensitive"
            >
                <InlineSVG
                    className="case-match"
                    element="i"
                    raw={false}
                    src="<svg></svg>"
                />
            </button>
            <button
                className="whole-word-btn"
                onClick={[Function]}
                title="Whole word"
            >
                <InlineSVG
                    className="whole-word-match"
                    element="i"
                    raw={false}
                    src="<svg></svg>"
                />
            </button>
        </div>
    </div>
</div>,
  "nodes": Array [
    <div
      className="search-bar"
>
      <SearchInput
            count={undefined}
            handleClose={[Function]}
            handleNext={[Function]}
            handlePrev={[Function]}
            onChange={[Function]}
            onKeyUp={[Function]}
            placeholder="Search in file…"
            query=""
            size=""
            summaryMsg=""
      />
      <div
            className="search-bottom-bar"
      >
            <div
                  className="search-modifiers"
            >
                  <button
                        className="regex-match-btn"
                        onClick={[Function]}
                        title="Regex"
                  >
                        <InlineSVG
                              className="regex-match"
                              element="i"
                              raw={false}
                              src="<svg></svg>"
                        />
                  </button>
                  <button
                        className="case-sensitive-btn"
                        onClick={[Function]}
                        title="Case sensitive"
                  >
                        <InlineSVG
                              className="case-match"
                              element="i"
                              raw={false}
                              src="<svg></svg>"
                        />
                  </button>
                  <button
                        className="whole-word-btn"
                        onClick={[Function]}
                        title="Whole word"
                  >
                        <InlineSVG
                              className="whole-word-match"
                              element="i"
                              raw={false}
                              src="<svg></svg>"
                        />
                  </button>
            </div>
      </div>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <SearchBar
        query=""
        searchOn={true}
        searchResults={Object {}}
        selectedResultIndex={0}
        selectedSymbolType="functions"
        symbolSearchOn={true}
        symbolSearchResults={Array []}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": SearchBar {
        "_reactInternalInstance": [Circular],
        "buildSummaryMsg": [Function],
        "clearSearch": [Function],
        "closeSearch": [Function],
        "context": Object {
          "shortcuts": undefined,
        },
        "doSearch": [Function],
        "onChange": [Function],
        "onEscape": [Function],
        "onKeyUp": [Function],
        "props": Object {
          "query": "",
          "searchOn": true,
          "searchResults": Object {},
          "selectedResultIndex": 0,
          "selectedSymbolType": "functions",
          "symbolSearchOn": true,
          "symbolSearchResults": Array [],
        },
        "refs": Object {},
        "renderSearchModifiers": [Function],
        "searchContents": [Function],
        "searchInput": [Function],
        "selectSearchInput": [Function],
        "setSearchValue": [Function],
        "state": Object {
          "count": 0,
          "index": -1,
          "selectedResultIndex": 0,
        },
        "toggleSearch": [Function],
        "traverseResults": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": Object {
        "_currentElement": <div
          className="search-bar"
>
          <SearchInput
                    count={undefined}
                    handleClose={[Function]}
                    handleNext={[Function]}
                    handlePrev={[Function]}
                    onChange={[Function]}
                    onKeyUp={[Function]}
                    placeholder="Search in file…"
                    query=""
                    size=""
                    summaryMsg=""
          />
          <div
                    className="search-bottom-bar"
          >
                    <div
                              className="search-modifiers"
                    >
                              <button
                                        className="regex-match-btn"
                                        onClick={[Function]}
                                        title="Regex"
                              >
                                        <InlineSVG
                                                  className="regex-match"
                                                  element="i"
                                                  raw={false}
                                                  src="<svg></svg>"
                                        />
                              </button>
                              <button
                                        className="case-sensitive-btn"
                                        onClick={[Function]}
                                        title="Case sensitive"
                              >
                                        <InlineSVG
                                                  className="case-match"
                                                  element="i"
                                                  raw={false}
                                                  src="<svg></svg>"
                                        />
                              </button>
                              <button
                                        className="whole-word-btn"
                                        onClick={[Function]}
                                        title="Whole word"
                              >
                                        <InlineSVG
                                                  className="whole-word-match"
                                                  element="i"
                                                  raw={false}
                                                  src="<svg></svg>"
                                        />
                              </button>
                    </div>
          </div>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="search-bar"
>
          <SearchInput
                    count={undefined}
                    handleClose={[Function]}
                    handleNext={[Function]}
                    handlePrev={[Function]}
                    onChange={[Function]}
                    onKeyUp={[Function]}
                    placeholder="Search in file…"
                    query=""
                    size=""
                    summaryMsg=""
          />
          <div
                    className="search-bottom-bar"
          >
                    <div
                              className="search-modifiers"
                    >
                              <button
                                        className="regex-match-btn"
                                        onClick={[Function]}
                                        title="Regex"
                              >
                                        <InlineSVG
                                                  className="regex-match"
                                                  element="i"
                                                  raw={false}
                                                  src="<svg></svg>"
                                        />
                              </button>
                              <button
                                        className="case-sensitive-btn"
                                        onClick={[Function]}
                                        title="Case sensitive"
                              >
                                        <InlineSVG
                                                  className="case-match"
                                                  element="i"
                                                  raw={false}
                                                  src="<svg></svg>"
                                        />
                              </button>
                              <button
                                        className="whole-word-btn"
                                        onClick={[Function]}
                                        title="Whole word"
                              >
                                        <InlineSVG
                                                  className="whole-word-match"
                                                  element="i"
                                                  raw={false}
                                                  src="<svg></svg>"
                                        />
                              </button>
                    </div>
          </div>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <SearchBar
    query=""
    searchOn={true}
    searchResults={Object {}}
    selectedResultIndex={0}
    selectedSymbolType="functions"
    symbolSearchOn={true}
    symbolSearchResults={Array []}
/>,
}
`;
